smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/UserTuneManager.java

```diff
@@ -0,0 +1,142 @@
+ *     XEP-0118: User Tune</a>
+ */
+public final class UserTuneManager extends Manager {
+
+    public static final String USERTUNE_NODE = "http://jabber.org/protocol/tune";
+    public static final String USERTUNE_NOTIFY = USERTUNE_NODE + "+notify";
+
+    private static final Map<XMPPConnection, UserTuneManager> INSTANCES = new WeakHashMap<>();
+
+    private final Set<UserTuneListener> userTuneListeners = new HashSet<>();
+    private final AsyncButOrdered<BareJid> asyncButOrdered = new AsyncButOrdered<>();
+    private final PubSubManager pubSubManager;
+    private final int WAIT_TIME_MILLI = 5000;
+    private long timeForLastPublish;
+
+    public static synchronized UserTuneManager getInstanceFor(XMPPConnection connection) throws NotLoggedInException {
+        UserTuneManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new UserTuneManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+    private UserTuneManager(XMPPConnection connection) throws NotLoggedInException {
+        super(connection);
+        PepManager.getInstanceFor(connection).addPepListener(new PepListener() {
+            @Override
+            public void eventReceived(EntityBareJid from, EventElement event, Message message) {
+                if (!USERTUNE_NODE.equals(event.getEvent().getNode())) {
+                    return;
+                }
+
+                final BareJid contact = from.asBareJid();
+                asyncButOrdered.performAsyncButOrdered(contact, new Runnable() {
+                    @Override
+                    public void run() {
+                        ItemsExtension items = (ItemsExtension) event.getExtensions().get(0);
+                        PayloadItem<?> payload = (PayloadItem) items.getItems().get(0);
+                        UserTuneElement tune = (UserTuneElement) payload.getPayload();
+
+                        for (UserTuneListener listener : userTuneListeners) {
+                            listener.onUserTuneUpdated(contact, message, tune);
+                        }
+                    }
+                });
+            }
+        });
+        ServiceDiscoveryManager.getInstanceFor(connection).addFeature(USERTUNE_NOTIFY);
+        pubSubManager = PubSubManager.getInstanceFor(getAuthenticatedConnectionOrThrow(), connection().getUser().asBareJid());
+
+    }
+
+    public void clearUserTune() throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+        publishUserTune(UserTuneElement.EMPTY_USER_TUNE);
+    }
+
+    public void publishUserTune(UserTuneElement userTuneElement) throws NotLoggedInException, NotALeafNodeException, NoResponseException, NotConnectedException, XMPPErrorException, InterruptedException {
+
+        LeafNode node = pubSubManager.getOrCreateLeafNode(USERTUNE_NODE);
```

Comment:
@Flowdalic I've looked at other XEP implementations based on PEP and every single one of them uses PubSub. It is alright to still go with PepManager instead of PubsubManager ?

Reply:
Which one? I'd say you should use `PepManager` whenever it is possible, which appears to be the case here. If `PepManager` does not provide the required functionality, then we should evaluate if we should put it there.

Reply:
@Flowdalic 
Xep-0080 and xep-0107  seem to be based on pep along with xep-0108 which has not been implemented yet.
[Geolocation](https://github.com/igniterealtime/Smack/blob/master/smack-extensions/src/main/java/org/jivesoftware/smackx/geoloc/GeoLocationManager.java)
[User Mood](https://github.com/igniterealtime/Smack/blob/master/smack-extensions/src/main/java/org/jivesoftware/smackx/mood/MoodManager.java )
 Please take a look at these and let me know what you think should be my approach. You might want me to give minor touching to these other modules as well 😉

Reply:
Yep, we should probably check if those shouldn't use `PepManager` instead. Possible that the code predates, at least some functionaly of, `PepManager`.

Reply:
There is now `PepManager.publish()` which this method should use.